libmscore/realizedharmony.cpp

```diff
@@ -0,0 +1,493 @@
+                  if (!(omit & (1 << 7))) {
+                        if (quality == "major")
+                              ret.insert(11 + RANK_MULT*RANK_7TH, tpcInterval(rootTpc, 7, 0));
+                        else if (quality == "diminished")
+                              ret.insert(9 + RANK_MULT*RANK_7TH, tpcInterval(rootTpc, 7, -2));
+                        else //dominant or augmented or minor
+                              ret.insert(10 + RANK_MULT*RANK_7TH, tpcInterval(rootTpc, 7, -1));
+                        }
+                  break;
+            case 6:
+                  if (!(omit & (1 << 6))) {
+                        ret.insert(9 + RANK_MULT*RANK_ADD, tpcInterval(rootTpc, 6, 0));     //maj6
+                        omit |= 1 << 13; //no need to add/alter 6 chords
+                        }
+                  break;
+            case 5:
+                  //omitted third already
+                  break;
+            case 4:
+                  if (!(omit & (1 << 4)))
+                        ret.insert(5 + RANK_MULT*RANK_ADD, tpcInterval(rootTpc, 4, 0));     //p4
+                  break;
+            case 2:
+                  if (!(omit & (1 << 2)))
+                        ret.insert(2 + RANK_MULT*RANK_ADD, tpcInterval(rootTpc, 2, 0));     //maj2
+                  omit |= 1 << 9; //make sure we don't add altered 9 when theres a 2
+                  break;
+            case 69:
+                  ret.insert(9 + RANK_MULT*RANK_ADD, tpcInterval(rootTpc, 6, 0));     //maj6
+                  ret.insert(2 + RANK_MULT*RANK_ADD, tpcInterval(rootTpc, 9, 0));     //maj6
+                  omit = ~0; //no additions/alterations for a 69 chord
+                  break;
+            default:
+                  break;
+            }
+
+      Harmony* next = _harmony->findNext();
+      if (!_literal && next) {
+            //jazz interpretation
+            QString qNext = next->parsedForm()->quality();
+            //pitch from current to next harmony normalized to a range between 0 and 12
+            //add PITCH_DELTA_OCTAVE before modulo so that we can ensure arithmetic mod rather than computer mod
+            int keyTpc = int(next->staff()->key(next->tick())) + 14; //tpc of key (ex. F# major would be Tpc::F_S)
+            int keyTpcMinor = keyTpc + 3; //tpc of relative minor of the key signature (ex. F minor would be Tpc::Ab + 3 = Tpc::F)a
+            int pitchBetween = (tpc2pitch(next->rootTpc()) + PITCH_DELTA_OCTAVE - tpc2pitch(rootTpc)) % PITCH_DELTA_OCTAVE;
+
+            //dont add 9 for chords with altered 5s
+            if (!(omit & (1 << 9)) && !alt5) {
+                  if (quality == "dominant" && pitchBetween == 5 && qNext == "minor"
+                      /*((next->rootTpc() == keyTpc && qNext == "major") || (next->rootTpc() == keyTpcMinor && qNext == "minor"))*/) {
+                        //flat 9 for dominant to a tonic chord a P4 up
+                        //only for minor chords for now
```

Comment:
FWIW, adding this for dominants resolving to major sevenths (eg, V7 resolving to I, or I7 resolving to IV) would be my main suggestion for improvement in the note choice algorithm.